### ISSUE TITLE
No support for X-Forwarded-Port

### ISSUE THREAD

**Issue by Lanoli** [aspnet/BasicMiddleware#204]
X-Forwarded-Port header is not supported, only ports in X-Forwarded-Host

**Comment by Tratcher** [comment]
Is the port in the Host inadequate / inaccurate?

**Comment by Lanoli** [comment]
It works, but I have an existing environment that uses the X-Forwarded-Port header to communicate the port instead of putting it in the X-Forwarded-Host header. Both methods seem to be valid.

**Comment by Abdruggi** [comment]
Libraries like Swashbuckle expect the port in a separate X-Forwarded-Port header. On the other side, AWS Load Balancing sets the Port header (http://docs.aws.amazon.com/elasticloadbalancing/latest/classic/x-forwarded-headers.html#x-forwarded-port). If the task is still open i would be willing to implement it.

**Comment by Tratcher** [comment]
@Abdruggi does the x-forwarded-port value differ from the value included in the Host header?

You're welcome to submit a PR for this.

**Comment by Abdruggi** [comment]
The x-forwarded-host header should not include a port if the x-forwarded-port header is set. I would suggest something like this: 

If just the x-forwarded-port header is present:
- request.Host.Port = x-forwarded-port header 

If just the x-forwarded-host header is present:
- request.Host.Host = x-forwarded-host header (don’t strip the port if available)

if both headers are present: 
- request.Host.Port = x-forwarded-port header 
- request.Host.Host = x-forwarded-host header (strip the port if available)

What do you think about this?

**Comment by elevate-andrewlock** [comment]
I've just run into this too, only the `x-forwarded-port` header is forwarded in my setup (and `X-Forwarded-Proto`, `X-Forwarded-For`, but no `x-forwarded-host`)

**Comment by biaol-odata** [comment]
@Tratcher @Abdruggi Any timeline information available for resolution of this issue? If this is not going to be resolved in near future, we will probably consider alternative of patching it up in upper layer (WebApi for OData) to unblock some user scenario. Thanks.

**Comment by Tratcher** [comment]
No, we don't have any plans for this feature at the moment. What proxy is sending you that header?

**Comment by biaol-odata** [comment]
@elevate-andrewlock ?

**Comment by andrewlock** [comment]
Both the NGINX and HAProxy setups we currently have send that `x-forwarded-port`. I believe AWS ELBs also send this header. NGINX also sends the `x-forwarded-host` (without the port), but HAProxy does not. Consequently, we have not been able to use the `XForwardedHost` header. 

As I workaround, I've basically had to duplicate the `ForwardedHeadersMiddleware` in miniature, but it really feels like @Abdruggi's [comment](https://github.com/aspnet/BasicMiddleware/issues/204#issuecomment-320990099) would be the most logical treatment of this header.

**Comment by Tratcher** [comment]
@shirhatti

**Comment by shirhatti** [comment]
Clearing Backlog milestone so we can re-triage and assess priority

**Comment by muratg** [comment]
@shirhatti Can you take a look?

**Comment by shirhatti** [comment]
@Abdruggi @andrewlock  I spent some time taking a look at your suggestions. Unless I'm mistaken I couldn't find any RFC that defines the behavior that you've called out?

We'd be extremely hesitant to introduce non-standard behavior.

**Comment by andrewlock** [comment]
@shirhatti, no as far as I can see, there is no RFC governing X-Forwarded-Port. However, while not standard, it is extremely common. For example [AWS use X-Forwarded-Port in the Elastic Load Balancers](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/x-forwarded-headers.html), as does [the standard Kubernetes NGINX ingress-controller](https://github.com/kubernetes/ingress-nginx/blob/master/rootfs/etc/nginx/template/nginx.tmpl#L1074) (which is where I found the issue). Even Swashbuckle [uses the header to correctly generate links](https://github.com/domaindrivendev/Swashbuckle/blob/master/Swashbuckle.Core/Application/SwaggerDocsConfig.cs#L308).

I don't feel like supporting `X-Forwarded-Port` would be a particularly controversial requirement given its widespread use, and the fact that it can be easily (optionally) added to existing behaviour. I think [This comment](https://github.com/aspnet/BasicMiddleware/issues/204#issuecomment-320990099) best describes the expected behaviour, which is to simply use the forwarded port if it's provided. We've already seen bugs in your own products (OData) which were highlighted ("caused" is probably a bit strong) by this issue. 

I understand the reticence to add "non standard" behaviour, but it seems like this will be something that bites people without them realising, and can cause some tricky bugs to identify. What about that "pit of success" 😉. Obviously this is mostly selfish - I really don't want to have to maintain my own version of the forwarded ports middleware (as I'm currently doing locally), but if you commit to _not_ implementing this I'll just publish my version on NuGet.

**Comment by shirhatti** [comment]
> I understand the reticence to add "non standard" behaviour, but it seems like this will be something that bites people without them realising, and can cause some tricky bugs to identify.

I would love to understand this scenario better. Given that the `X-Forwarded-Host` already contains port, I'm missing these pits of failures without `X-Forwarded-Port`.

**Comment by andrewlock** [comment]
Sure, so for the examples I've described (and experienced):

* AWS ELB doesn't use `X-Forwarded-Host` at all, only `X-Forwarded-Port`. This is probably the most common occurence (they send `X-Forwarded-Port`, `X-Forwarded-Proto`, `X-Forwarded-For`, but no `X-forwarded-host`). 
* Some upstream proxies won't set the port in the `X-Forwarded-Host` - instead `X-Forwarded-Host` contains the hostname only, and `X-Forwarded-Port` contains the port. I can't remember the exact setup where I saw this, but I believe it was behind an HAProxy instance.

**Comment by Tratcher** [comment]
What does ELB put in the real Host header? The back-end machine name and port? Or the original public value?

**Comment by shirhatti** [comment]
As I best I can tell from ELB docs, it sends the original unmodified Host Header.

For HAProxy, there is no default behavior. The suggested configuration has this
```
  http-request set-header X-Forwarded-Host %[req.hdr(host)]
  http-request set-header X-Forwarded-Port %[dst_port]
  http-request set-header X-Forwarded-Proto http if !{ ssl_fc }
  http-request set-header X-Forwarded-Proto https if { ssl_fc }
```

From this it's evident that that `X-Forwarded-Host` still contains the port information.

**Comment by shirhatti** [comment]
If anything, this thread is testament to why we don't like to ship non-standard behavior 😄 

I'm going to backlog this for now, unless we have evidence that there seems to be consensus among what different proxies do.
